### PR TITLE
troubleshooting/bootkube_recovery_tool.md: add clean up steps

### PR DESCRIPTION
**Pull request by radhikapc**
ref: https://github.com/coreos/tectonic-installer/issues/1428

### CODE REVIEW

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 1]
```diff
@@ -27,11 +27,38 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
 
 Use the `bootkube recover` command to re-bootstrap the self-hosted control plane in the event of partial or total self-hosted control plane loss. `bootkube recover` does not recover a cluster directly. Instead, it extracts the control plane configuration from an available source and renders manifests in a format that the `bootkube start` command can use to reboot the cluster.
 
-For best results use the latest version of `bootkube` when using the `bootkube recover` command, regardless of which version was used to create the cluster. To see available options, run:
+Use the latest version of `bootkube` when using the `bootkube recover` command, regardless of which version was used to create the cluster. To see available options, run:
 
 `bootkube recover --help`
 
-To recover a cluster, perform the following:
+### Cleaning up the existing etcd resources
```
Comment (xiang90) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: cleaning up the existing Kubernetes state 

not just etcd. it is actually more k8s related state.

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 2]
```diff
@@ -27,11 +27,38 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+    ```
+    docker kill $(docker ps -q)
+    docker rm $(docker ps -a -q)
+    ```
+
+3. Remove the active Kubernetes active manifest:
+
+    `rm /etc/kubernetes/manifests/*`
+
+4. If self-hosted etcd is enabled, remove associated data:
```
Comment (xiang90) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: we probably want to warn people a little bit on this. maybe ask them to backup etcd data before removing?

**Comment by xiang90** [comment]
looks good to me in general.

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 3]
```diff
@@ -27,11 +27,38 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+
+    `rm -rf /var/etcd/kube-*`
+
+5. Restart kubelet service:
+
+    `systemctl restart kubectl`
```
Comment (xiang90) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: systemctl restart kubelet.service?

**Comment by estroz** [comment]
ok to test

**Comment by estroz** [comment]
ok to test

**Comment by xiang90** [comment]
lgtm

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 4]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
 `bootkube recover --help`
 
-To recover a cluster, perform the following:
+### Cleaning up Kubernetes resources
+
+Before recovering a cluster, clean up the existing Kubernetes resources. To do so, run the following on all the master nodes.
+
+1. Stop the kubelet service:
+
+   `systemctl stop kubelet.service`
```
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: found that this needed to be sudo

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 5]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+
+2. Remove the docker state:
+
+    ```
+    docker kill $(docker ps -q)
```
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: `$` in front of all of these

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 6]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+    docker rm $(docker ps -a -q)
+    ```
+
+3. Remove the active Kubernetes active manifest:
+
+    `rm -rf /etc/kubernetes/manifests/*`
```
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: found that this needed to be sudo

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 7]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+
+4. Remove any trace of pods (logs, state, configuration) that the kubelet was aware of:
+
+   `rm -rf /var/lib/kubelet/pods`
```
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: sudo

Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: Even with sudo, this did not work for me, spewed a bunch of:

```
rm: cannot remove '/var/lib/kubelet/pods/e3d123f3-6cc3-11e7-bb4f-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/1cc369d7-6728-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/1cc60359-6728-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/dfb5ccd6-6727-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/d6912990-6728-11e7-a172-02c7e4d4d0f8/volumes/kubernetes.io~secret/default-token-kbmm6': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/1cc47b6f-6728-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/1cc8c865-6728-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/1cc8c865-6728-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/secrets': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/dfb3a234-6727-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/dfb3a234-6727-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/secrets': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/e0491ac8-6728-11e7-a172-02c7e4d4d0f8/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/e0491ac8-6728-11e7-a172-02c7e4d4d0f8/volumes/kubernetes.io~secret/member-peer-tls': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/e0491ac8-6728-11e7-a172-02c7e4d4d0f8/volumes/kubernetes.io~secret/member-server-tls': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/e0491ac8-6728-11e7-a172-02c7e4d4d0f8/volumes/kubernetes.io~secret/etcd-client-tls': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/e351e704-6cc3-11e7-bb4f-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/d18f54d0-6728-11e7-a172-02c7e4d4d0f8/volumes/kubernetes.io~secret/default-token-kbmm6': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/0125571f-6729-11e7-a172-02c7e4d4d0f8/volumes/kubernetes.io~secret/default-token-kbmm6': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/1cc66376-6728-11e7-ada1-06c278b62e96/volumes/kubernetes.io~secret/default-token-krhns': Device or resource busy
```

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 8]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+
+5. If self-hosted etcd is enabled:
+
+    1. Back up the associated data.
+
+       `cp $(find /var/etcd/ -iname db) /root/backup.db`
```
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: Did not work:

```
core@ip-10-0-0-230 ~ $ cp $(find /var/etcd/ -iname db) /root/backup.db
find: `/var/etcd/kube-system-kube-etcd-0001': Permission denied
cp: missing destination file operand after '/root/backup.db'
Try 'cp --help' for more information.
```

Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: What did work:

```
sudo cp $(sudo find /var/etcd/ -iname db) /root/backup.db
```

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 9]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+
+    2. Remove the data from the master nodes:
+
+       `rm -rf /var/etcd/kube-*`
```
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: sudo

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 10]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+
+6. Restart the kubelet service:
+
+    `systemctl restart kubectl.service`
```
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: Needs sudo and the unit is called `kubelet.service`

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 11]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
...
+
+### Running bootkube
 
 1. Copy `bootkube` to a master node and then SSH to it:
```
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: Putting my user hat on...where do I get bootkube from?

Putting my PM hat on...we have requirements on the version of bootkube, right? Do we need to tell them to get a specific one?

File: Documentation/troubleshooting/bootkube_recovery_tool.md (hunk 7 above)
Comment (xiang90) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: is the kubelet and all containers stopped? this means the dir is still mounted somewhere.

**Comment by radhikapc** [comment]
@robszumski thanx for testing- Duffie also going to test these steps, and is underway.

File: Documentation/troubleshooting/bootkube_recovery_tool.md (hunk 11 above)
Comment (radhikapc) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: Won't Tectonic Installer by default install it ?

File: Documentation/troubleshooting/bootkube_recovery_tool.md (hunk 10 above)
Comment (radhikapc) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: Oops that was a typo

**Comment by robszumski** [comment]
Here's what did work for me:

```
core@ip-10-0-0-230 ~ $ curl -L https://github.com/kubernetes-incubator/bootkube/releases/download/v0.5.1/bootkube.tar.gz > bootkube.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   605    0   605    0     0   2952      0 --:--:-- --:--:-- --:--:--  3009
100 38.0M  100 38.0M    0     0  2767k      0  0:00:14  0:00:14 --:--:-- 1833k
core@ip-10-0-0-230 ~ $ tar -xvzf  bootkube.tar.gz
bin/linux/bootkube
bin/darwin/bootkube
bin/linux/checkpoint
core@ip-10-0-0-230 ~ $ ls
bin  bootkube.tar.gz
core@ip-10-0-0-230 ~ $ cd bin/linux/
core@ip-10-0-0-230 ~ $ sudo chown core:core /root/backup.db
core@ip-10-0-0-230 ~/bin/linux $ ./bootkube recover --recovery-dir=/home/core/recovered --etcd-backup-file=/root/backup.db --kubeconfig=/etc/kubernetes/
ca.crt               checkpoint-secrets/  inactive-manifests/  kubelet.env
checkpoint-iptables/ cni/                 kubeconfig           manifests/
core@ip-10-0-0-230 ~/bin/linux $ ./bootkube recover --recovery-dir=/home/core/recovered --etcd-backup-file=/root/backup.db --kubeco
nfig=/etc/kubernetes/kubeconfig
Attempting recovery using etcd backup file at "/root/backup.db"...
Writing asset: /etc/kubernetes/manifests/recovery-etcd.yaml
Error: open /etc/kubernetes/manifests/recovery-etcd.yaml: permission denied
open /etc/kubernetes/manifests/recovery-etcd.yaml: permission denied
core@ip-10-0-0-230 ~/bin/linux $ sudo !!
sudo ./bootkube recover --recovery-dir=/home/core/recovered --etcd-backup-file=/root/backup.db --kubeconfig=/etc/kubernetes/kubeconfig
Attempting recovery using etcd backup file at "/root/backup.db"...
Writing asset: /etc/kubernetes/manifests/recovery-etcd.yaml
Waiting for etcd server to start...
Writing asset: /home/core/recovered/bootstrap-manifests/bootstrap-kube-apiserver.yaml
Writing asset: /home/core/recovered/bootstrap-manifests/bootstrap-kube-controller-manager.yaml
Writing asset: /home/core/recovered/bootstrap-manifests/bootstrap-kube-scheduler.yaml
Writing asset: /home/core/recovered/tls/etcd-client-ca.crt
Writing asset: /home/core/recovered/tls/etcd-client.crt
Writing asset: /home/core/recovered/tls/etcd-client.key
Writing asset: /home/core/recovered/tls/etcd/peer-ca.crt
Writing asset: /home/core/recovered/tls/etcd/peer.crt
Writing asset: /home/core/recovered/tls/etcd/peer.key
Writing asset: /home/core/recovered/tls/etcd/server-ca.crt
Writing asset: /home/core/recovered/tls/etcd/server.crt
Writing asset: /home/core/recovered/tls/etcd/server.key
Writing asset: /home/core/recovered/tls/secrets/kube-apiserver/etcd-client-ca.crt
Writing asset: /home/core/recovered/tls/secrets/kube-apiserver/etcd-client.crt
Writing asset: /home/core/recovered/tls/secrets/kube-apiserver/etcd-client.key
Writing asset: /home/core/recovered/tls/secrets/kube-apiserver/service-account.pub
Writing asset: /home/core/recovered/tls/secrets/kube-apiserver/apiserver.crt
Writing asset: /home/core/recovered/tls/secrets/kube-apiserver/apiserver.key
Writing asset: /home/core/recovered/tls/secrets/kube-apiserver/ca.crt
Writing asset: /home/core/recovered/tls/secrets/kube-controller-manager/ca.crt
Writing asset: /home/core/recovered/tls/secrets/kube-controller-manager/service-account.key
Writing asset: /home/core/recovered/bootstrap-manifests/bootstrap-etcd.yaml
Writing asset: /home/core/recovered/etcd/bootstrap-etcd-service.json
Writing asset: /home/core/recovered/etcd/migrate-etcd-cluster.json
Writing asset: /home/core/recovered/auth/kubeconfig
core@ip-10-0-0-230 ~/bin/linux $ ./bootkube start --asset-dir=/home/core/recovered
Starting temporary bootstrap control plane...
Error: mkdir /etc/kubernetes/bootstrap-secrets: permission denied
Tearing down temporary bootstrap control plane...
Error: mkdir /etc/kubernetes/bootstrap-secrets: permission denied
Error: mkdir /etc/kubernetes/bootstrap-secrets: permission denied
mkdir /etc/kubernetes/bootstrap-secrets: permission denied
core@ip-10-0-0-230 ~/bin/linux $ sudo !!
sudo ./bootkube start --asset-dir=/home/core/recovered
Starting temporary bootstrap control plane...
...a whole bunch of stuff...
I0720 05:03:31.020354   30476 migrate.go:91] removed boot-etcd from the etcd cluster
Tearing down temporary bootstrap control plane...
```

Note a few things:
 - write to `/etc` requires sudo, as covered any times above
 - found the answer from above about `recovered`, its created for you
 - I referenced the kubeconfig already on disk
 - seems nuts that we can't ship a script that does this for you. maybe not the etcd backup part, but the actual recovery seems to have defined paths for where the kubeconfig is, we could standardize the etcd backup location, recovered directory location, etc.

@aaronlevy can we ship a script for this and write it on the masters? it would have turned this 1hour ordeal into a super easy thing...

File: Documentation/troubleshooting/bootkube_recovery_tool.md (hunk 7 above)
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: saw this too late, i already recovered it (guess it didnt matter too much)

File: Documentation/troubleshooting/bootkube_recovery_tool.md (hunk 11 above)
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: I could not find it hunting around on disk on the master. But if it was, we would still need to tell them where it was.

Plus, your instructions say copy it from somewhere to the master. It's not in the install download or the assets bundle.

**Comment by xiang90** [comment]
@robszumski 

if you run recovery on a failed machine, you need the cleanup steps. and this is what most people will hit.

if you run recovery on a fresh machine, then you do not need the cleanup steps.

File: Documentation/troubleshooting/bootkube_recovery_tool.md (hunk 7 above)
Comment (xiang90) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: @robszumski 

it matters when self hosted etcd is enabled. kubelet has a bug to mount previous used empty dir to static pod if you do not clean this up.

File: Documentation/troubleshooting/bootkube_recovery_tool.md (hunk 11 above)
Comment (radhikapc) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: I made a wrong assumption by thinking people would know it already! My mistake.

File: Documentation/troubleshooting/bootkube_recovery_tool.md (hunk 7 above)
Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: I can't get these to remove. Verified that both docker and the kubelet are stopped.

Comment (xiang90) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: /cc @coresolve

Comment (xiang90) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: probably try lsof +D xxx to check who opens these files. but i will let @coresolve to provide more details on this.

Comment (robszumski) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: `lsof +D` doesn't help either...

File: Documentation/troubleshooting/bootkube_recovery_tool.md [hunk 12]
```diff
@@ -27,11 +29,48 @@ The pod checkpointer is enabled by default and is highly recommended for all clu
 
 Use the `bootkube recover` command to re-bootstrap the self-hosted control plane in the event of partial or total self-hosted control plane loss. `bootkube recover` does not recover a cluster directly. Instead, it extracts the control plane configuration from an available source and renders manifests in a format that the `bootkube start` command can use to reboot the cluster.
 
-For best results use the latest version of `bootkube` when using the `bootkube recover` command, regardless of which version was used to create the cluster. To see available options, run:
+Use the latest version of `bootkube` when using the `bootkube recover` command, regardless of which version was used to create the cluster. To see available options, run:
 
 `bootkube recover --help`
 
-To recover a cluster, perform the following:
+### Cleaning up Kubernetes resources
```
Comment (xiang90) [inline Documentation/troubleshooting/bootkube_recovery_tool.md]: this should be put before bootkube recover

**Comment by radhikapc** [comment]
@xiang90 why would someone run recovery on a fresh machine ? I am still unclear about recovering on a fresh machine vs recovering from a failed machine (=> one of the control plane component is down).

**Comment by aaronlevy** [comment]
@robszumski not necessarily opposed to adding a script as part of installation step - but it could get stale if we're just installing it to all the hosts (and doesn't help recovery on fresh machines). I'd probably lean toward linking to any scripts in the recovery docs instead (to keep fresh, allow easy access from clean machines).

Some of these steps could also be included in the actual recovery command itself (to a degree). We have an open issue in bootkube to consider some of the existing-state cleanup steps being added: https://github.com/kubernetes-incubator/bootkube/issues/651

**Comment by radhikapc** [comment]
https://github.com/kubernetes-incubator/bootkube/issues/658 blocks this PR

**Comment by robszumski** [comment]
This is blocked on a few things, but it's better than what we have live. Anyone against merging as-is?

**Comment by xiang90** [comment]
@robszumski I am fine with merging as it is.

**Comment by robszumski** [comment]
Just docs, merging

**Pull request merged**